### PR TITLE
Request Parser

### LINKED ISSUE #10: Http parser

### PR DESCRIPTION
**Pull request by maartenRamadan**
#32

### CODE REVIEW

File: src/server/http/HTTPParser.cpp [hunk 1]
```diff
@@ -3,10 +3,268 @@
...
+	_methodMap["INVALID"] = INVALID;
+	_methodMap["GET"] = GET;
+	_methodMap["HEAD"] = HEAD;
+	_methodMap["POST"] = POST;
+	_methodMap["PUT"] = PUT;
+	_methodMap["DELETE"] = DELETE;
+	_methodMap["CONNECT"] = CONNECT;
+	_methodMap["OPTIONS"] = OPTIONS;
+	_methodMap["TRACE"] = TRACE;
+}
```
Comment (pmerku) [inline src/server/http/HTTPParser.cpp]: You can use `utils::CreateMap()` to initialize the map, check how I do it in the ResponseBuilder.cpp

**Review by pmerku — CHANGES_REQUESTED** [review]
Add a rule for `.DS_Store` in the .gitignore and remove the files from your fork

File: src/server/http/HTTPParser.cpp [hunk 2]
```diff
@@ -3,10 +3,268 @@
...
+				return "PUT";
+			case DELETE:
+				return "DELETE";
+			case CONNECT:
+				return "CONNECT";
+			case OPTIONS:
+				return "OPTIONS";
+			case TRACE:
+				return "TRACE";
+		}
```
Comment (pmerku) [inline src/server/http/HTTPParser.cpp]: To add a `default:` to the switch, to fix the compilation error

File: src/server/handlers/StandardHandler.cpp [hunk 3]
```diff
@@ -40,6 +40,7 @@ void StandardHandler::read(HTTPClient &client) {
 		client.connectionState = WRITING;
 		_eventBus->postEvent(ServerEventBus::CLIENT_STATE_UPDATED);
 	}
+	std::cout << client.data.request << std::endl; //
```
Comment (mrjvs) [inline src/server/handlers/StandardHandler.cpp]: its best to keep debug logging out of pr's (or use the logger with DEBUG flag)

File: src/server/http/HTTPParser.cpp (hunk 2 above)
Comment (mrjvs) [inline src/server/http/HTTPParser.cpp]: Use a static const map for easy and fast searching. the main branch has a CreateMap() templated builder that you can use to make it. (one map for enum to string, another one for string to enum). then use map.find() to find your value.

File: src/server/http/HTTPParser.cpp [hunk 4]
```diff
@@ -3,10 +3,268 @@
...
+					std::cout << "Header: [" << it->first << ": " << it->second << "]" << std::endl;
+			}
+			else
+				std::cout << "-NO HEADERS-" << std::endl;
+			if (x._body.length()) {
+				std::cout << "Body length: " << x._body.length() << std::endl;
+				std::cout << std::endl << "-BODY-" << std::endl << x._body << std::endl;
+			}
+			else
+				std::cout << std::endl << "-NO BODY-" << std::endl;
```
Comment (mrjvs) [inline src/server/http/HTTPParser.cpp]: dont use std::cout in ostream overload. use the `o` variable

File: src/server/http/HTTPParser.cpp (hunk 1 above)
Comment (mrjvs) [reply src/server/http/HTTPParser.cpp]: also make it static const :)

File: src/server/http/HTTPParser.cpp [hunk 5]
```diff
@@ -3,10 +3,268 @@
...
+	_R._method = _methodMap.find(parts[0])->second;
+
+	// CHECK URI
+	if (parts[1][0] != '/') {
+		_R._statusCode = 401;
+		globalLogger.logItem(logger::ERROR, "Invalid URI");
+		return ERROR;
+	}
+
+	std::string allowedURIChars("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.~!#$&'()*+,/:;=?@[]");
```
Comment (mrjvs) [inline src/server/http/HTTPParser.cpp]: make this a static const on the class.

File: src/utils/isDigit.cpp [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#include "utils/isDigit.hpp"
+
+int		utils::isDigit(int c)
+{
+	if (c >= '0' && c <='9')
+		return (1);
+	else
+		return (0);
```
Comment (mrjvs) [inline src/utils/isDigit.cpp]: shorten to `return (c >= '0' && c <='9');`

**Review by mrjvs — CHANGES_REQUESTED** [review]
also remove script.js from the repo for now (and maybe put in .gitignore). we'll add it properly when we add proper testing

File: src/includes/server/http/HTTPClientData.hpp [hunk 7]
```diff
@@ -5,27 +5,51 @@
 #ifndef HTTPCLIENTDATA_HPP
 #define HTTPCLIENTDATA_HPP
 
+#include "ServerTypes.hpp"
```
Comment (pmerku) [inline src/includes/server/http/HTTPClientData.hpp]:
```suggestion
#include "server/ServerTypes.hpp"
```

File: src/server/http/HTTPParser.cpp [hunk 8]
```diff
@@ -1,12 +1,258 @@
...
+			return ERROR;
+		}
+	}
+
+	// Set URI
+	_R._uri = parts[1];
+
+	// CHECK PROTOCOL
+	if (parts[2] != "HTTP/1.1") {
+		globalLogger.logItem(logger::ERROR, "Invalid protocol");
```
Comment (pmerku) [inline src/server/http/HTTPParser.cpp:225]:
```
_R._statusCode = 505;
```
Add the error status code

File: src/server/http/HTTPParser.cpp [hunk 9]
```diff
@@ -1,12 +1,258 @@
...
+		return ERROR;
+	}
+
+	// Append body
+	_R._body += rawRequest.substr(SOB, chunkSize);
+
+	// Check if body is complete
+	if (EOB != std::string::npos) // 0x0?
+	{
+		globalLogger.logItem(logger::INFO, "Succesfully parsed chunked body");
```
Comment (pmerku) [inline src/server/http/HTTPParser.cpp]: I would change this to DEBUG mode

File: src/server/http/HTTPParser.cpp [hunk 10]
```diff
@@ -1,12 +1,258 @@
...
+		return READY_FOR_WRITE;
+	}
+	else
+	{
+		globalLogger.logItem(logger::INFO, "UNFINISHED");
```
Comment (pmerku) [inline src/server/http/HTTPParser.cpp]: This one too

File: src/server/http/HTTPResponder.cpp [hunk 11]
```diff
@@ -29,15 +29,45 @@ void HTTPResponder::generateResponse(HTTPClient &client) {
...
+void	cgi::runCGI(HTTPClientRequest& request, const std::string& path) {
+	int pid;
+	char** args;
+
+	setMetaVars(request);
+	setEnv();
+
+	pid = fork();
+	if (pid < 0)
+		return ; // error
```
Comment (pmerku) [inline src/server/http/HTTPResponder.cpp]: Add the globalLogger with ERROR

Comment (pmerku) [reply src/server/http/HTTPResponder.cpp]: And add a TODO so we don't forget to check for proper cleanup and/or make an error response for the client later on

File: src/server/http/HTTPResponder.cpp [hunk 12]
```diff
@@ -29,15 +29,45 @@ void HTTPResponder::generateResponse(HTTPClient &client) {
...
+	if (!pid)
+	{
+		if (execve(args[0], args, _env) < 0)
+			return ; // error
```
Comment (pmerku) [inline src/server/http/HTTPResponder.cpp]: Same as previous comment

File: src/server/http/HTTPResponder.cpp [hunk 13]
```diff
@@ -29,15 +29,45 @@ void HTTPResponder::generateResponse(HTTPClient &client) {
...
 		CGIenv::env envp;
 		envp.setEnv(CGIenv::ENVBuilder()
-				.AUTH_TYPE("authentication")
-				.CONTENT_LENGTH("3000")
-				.CONTENT_TYPE("text/json")
-				.REQUEST_URI("/path")
-				.build()
-				);
+			.AUTH_TYPE(request._headers["AUTHORIZATION"]);
+			.CONTENT_LENGTH(utils::itos(request._body.length()));
```
Comment (pmerku) [inline src/server/http/HTTPResponder.cpp]: You can change `utils::itos` to `utils::intToString`, no need to make another function

File: src/includes/utils/itos.hpp [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#ifndef ITOS_HPP
+#define ITOS_HPP
+
+# include <string>
+
+namespace utils {
+	std::string		itos(int nb);
+}
+
+#endif //ITOS_HPP
```
Comment (pmerku) [inline src/includes/utils/itos.hpp]: Not needed

File: src/utils/itos.cpp [hunk 15]
```diff
@@ -0,0 +1,8 @@
+#include "utils/itos.hpp"
+
+std::string		utils::itos(int nb)
+{
+	if (nb == 0)
+		return "";
+	return itos(nb / 10) + (char)(nb % 10 + '0');
+}
```
Comment (pmerku) [inline src/utils/itos.cpp]: Not needed

**Review by pmerku — CHANGES_REQUESTED** [review]

File: makefile [hunk 16]
```diff
@@ -58,6 +58,13 @@ SRC	=\
 	utils/intToString.cpp\
 	utils/isSpace.cpp\
 	utils/mutex.cpp\
+	utils/split.cpp\
+	utils/atoi.cpp\
+	utils/stoi.cpp\
```
Comment (pmerku) [inline makefile:63]: Delete file

File: makefile [hunk 17]
```diff
@@ -82,50 +89,57 @@ HEADERS	=\
...
-    utils/AThread.hpp\
-    utils/CreateMap.hpp\
-    utils/CreateVector.hpp\
-    utils/ErrorThrow.hpp\
-    utils/intToString.hpp\
+  utils/AThread.hpp\
+	utils/split.hpp\
+	utils/atoi.hpp\
+	utils/toUpper.hpp\
+	utils/stoi.hpp\
```
Comment (pmerku) [inline makefile:113]: Delete file

File: script.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    //await sleep(1);
+	//send('9\r\n123456789\r\n');
+	//sleep(1);
+    //send('8\r\nBODYBODY\r\n');
+})
+
+s.connect(8081);
+
+
+//send('POST / HTTP/1.1\r\nHost: localhost:8080\r\nConnection: keep-alive\r\nCONTENT-LENGTH: 12\r\nUser-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36\r\nAccept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7\r\n\r\nBODYBODYBODY\r\n');
```
Comment (pmerku) [inline script.js]: Remove this file for now, like Jelle said

File: src/includes/utils/stoi.hpp [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+#define STOI_HPP
+
+# include <string>
+# include "isDigit.hpp"
+
+namespace utils {
+	size_t		stoi(const std::string& str);
+}
+
+#endif //STOI_HPP
```
Comment (pmerku) [inline src/includes/utils/stoi.hpp:11]: Remove file

File: src/utils/stoi.cpp [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+size_t		utils::stoi(const std::string& str)
+{
+	size_t ret = 0;
+	for (size_t i = 0; str[i]; ++i) {
+		if (!utils::isDigit(str[i])) // TODO
+			return (0);
+		ret = ret * 10 + (str[i] - '0');
+	}
+	return ret;
+}
```
Comment (pmerku) [inline src/utils/stoi.cpp:12]: Remove file

**Review by pmerku — CHANGES_REQUESTED** [review]

File: makefile (hunk 17 above)
Comment (maartenRamadan) [reply makefile:113]: What file?

File: script.js (hunk 18 above)
Comment (maartenRamadan) [reply script.js]: I did...

File: src/includes/utils/stoi.hpp (hunk 19 above)
Comment (maartenRamadan) [reply src/includes/utils/stoi.hpp:11]: it was itos that had to be removed right?

File: makefile (hunk 17 above)
Comment (pmerku) [reply makefile:113]: the stoi.hpp

Comment (pmerku) [reply makefile:113]: oh no sorry, brainfart

**Review by pmerku — APPROVED** [review]

File: script.js (hunk 18 above)
Comment (pmerku) [reply script.js]: It's still in the repo

**Review by mrjvs — APPROVED** [review]

**Pull request merged**